### PR TITLE
adding queue implementation[@Noran,@ABadr]

### PR DESCRIPTION
**Pull request by noran97**

### CODE REVIEW

File: conf.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+#define QUEUE_LINKED  (1)
+#define QUEUE_ARRAY   (2)
+
+//if you want queue array ,replace -QUEUE_ARRAY- by -QUEUE_LINKED-.
+#define QUEUE_TYPE     QUEUE_LINKED
+
+#if QUEUE_TYPE == QUEUE_ARRAY
+//if you want to change the size ,replace 100 with your size.
+#define arraysize    (100)
```
Comment (AbdelrahmanBadr22) [inline conf.h:12]: always make macros name uppercase separated by `_` character e.g. `ARRAY_SIZE`

File: conf.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+#endif
```
Comment (AbdelrahmanBadr22) [inline conf.h:13]: its preferred to write a comment beside `#endif` to make `#if` clear e.g. `#endif //QUEUE_TYPE`

File: conf.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#ifndef CONF_H_INCLUDED
```
Comment (AbdelrahmanBadr22) [inline conf.h:1]: rename this file `Queue_Cong.h`

File: queu.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#ifndef QUEU_H_INCLUDED
```
Comment (AbdelrahmanBadr22) [inline queu.h:1]: rename this file `Queue.h` to be compatible with other files

**Review by AbdelrahmanBadr22 — CHANGES_REQUESTED** [review]

File: Queue_Cong.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#ifndef CONF_H_INCLUDED
+#define CONF_H_INCLUDED
+
+#define QUEUE_LINKED  (1)
+#define QUEUE_ARRAY   (2)
```
Comment (AbdelrahmanBadr22) [inline Queue_Cong.h:5]: `Queue_Cong` !! 
make it `Queue_Config`

File: queue.c [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+#include "conf.h"
+#include "STD_TYPES.h"
+#include <assert.h>
+#if QUEUE_TYPE ==QUEUE_LINKED
+typedef struct node {
+    uint16_t key;
+    struct node* next;
+}node;
+
+typedef struct queue {
```
Comment (AbdelrahmanBadr22) [inline queue.c:12]: make this file named `Queue.c` ya noraaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaan

File: queue.c [hunk 7]
```diff
@@ -0,0 +1,111 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "conf.h"
```
Comment (AbdelrahmanBadr22) [inline queue.c:3]: اقسم بالله المفروض نعمل build ونتأكد انه بيعمل build قبل نعمل push عشان منبوظش اكواد الناس التانيه

**Review by AbdelrahmanBadr22 — CHANGES_REQUESTED** [review]

File: queue.c (hunk 6 above)
Comment (noran97) [reply queue.c:12]: شاهد شخص يحكم على الكتاب من عنوانه 
حاضر هغيره يا بشمهندس 👩‍🦯👩‍🦯

File: Queue.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+}queue_t;
+
+void queue_Initiate(queue_t*q);
+void queue_Push( queue_t *q,uint16_t val);
+uint16_t queue_Pop (queue_t*q);
+uint16_t queue_IsEmpty(queue_t*q);
+uint16_t queue_IsFull(queue_t*q);
+uint16_t queue_GetFront(queue_t *q);
+uint16_t queue_GetBack(queue_t*q);
+uint16_t queue_GetSize(queue_t*q);
```
Comment (AbdelrahmanBadr22) [inline Queue.h:37]: Prototypes of function for both implementations are same so no need to guard with `#if` or #else` and make the first character uppercase as follows `Queue_Initiate()` `Queue_Push` ...

File: Queue.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    uint16_t size;
+    struct node *front, *rear;
+}queue_t;
+
+void queue_Initialize(queue_t*q);
+void queue_Enqueue(queue_t* q, uint16_t val);
+uint16_t queue_Dequeue(queue_t* q);
+uint16_t queue_GetSize(queue_t* q);
+
+ #else
```
Comment (AbdelrahmanBadr22) [inline Queue.h:22]: `#elif ...` will be better as in case of error no functions exists

File: Queue_Config.h [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+#define QUEUE_LINKED  (1)
+#define QUEUE_ARRAY   (2)
+
+//if you want queue array ,replace -QUEUE_ARRAY- by -QUEUE_LINKED-.
+#define QUEUE_TYPE     QUEUE_LINKED
+
+#if QUEUE_TYPE == QUEUE_ARRAY
+//if you want to change the size ,replace 100 with your size.
+#define Array_Size    (100)
```
Comment (AbdelrahmanBadr22) [inline Queue_Config.h:12]: it's preferred to make all macros in uppercase e.g. `#define ARRAY_SIZE`

File: Queue_Config.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+#endif //QUEUE_TYPE != QUEUE_ARRAY
+#if QUEUE_TYPE != QUEUE_ARRAY  && QUEUE_TYPE != QUEUE_LINKED
+
+#error "Error in defining QUEUE_TYPE "
```
Comment (AbdelrahmanBadr22) [inline Queue_Config.h:16]: it's preferred in error message to refer to file name e.g. `"ERROR IN DEFINING QUEUE_TYPE @Queue_Config.h`

File: Queue.h [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+typedef struct queue {
+    uint16_t size;
+    struct node *front, *rear;
+}queue_t;
+
+#elif QUEUE_TYPE == QUEUE_ARRAY
+typedef struct queuearray
+{
+uint16_t size;
+uint16_t arr[ARRAY_SIZE];
```
Comment (AbdelrahmanBadr22) [inline Queue.h:21]: make the type of array as a macro and make the user determine which type he want e.g. `typedef   QueueType uint16`

File: Queue.h [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+uint16_t i;
```
Comment (AbdelrahmanBadr22) [inline Queue.h:22]: `i` really ? 
what is the meaning of that ???

File: Queue.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+}queue_t;
```
Comment (AbdelrahmanBadr22) [inline Queue.h:23]: make it compatible with other code e.g. `Queue_t`

**Pull request merged**